proxy/src/main/java/com/velocitypowered/proxy/connection/client/ConnectedPlayer.java

```diff
@@ -554,6 +558,21 @@ public void disconnect(Component reason) {
     }
   }
 
+  /**
+   * Checks and kicks the player when chat rate limit threshold has been reached.
+   *
+   * @return whether the message should be processed or not
+   */
+  public boolean isChatRateLimited() {
+    boolean allowed = limiter.attempt(this);
+
+    if (!allowed) {
+      disconnect(Component.translatable("disconnect.spam", NamedTextColor.RED));
+    }
+
+    return !allowed;
+  }
```

Comment:
This is extremely deceptive - a function that has a non-obvious side effect when called. I'd rather have the disconnect moved into the sites where `isChatRateLimited()` is called.